STYLEGUIDE.adoc

```diff
@@ -0,0 +1,426 @@
+= Plutus Haskell style guide
+
+The purpose of this document is to codify various aspects of how we write Haskell code as we contribute to Plutus. 
+
+The goal is twofold:
+To provide guidance in cases where it is not obvious what to do, or the choice is arbitrary.
+To help resolve disagreements between team members. Ideally these can be settled by referring to a recommendation in this document, or if not, then perhaps the resolution of the disagreement will create a new recommendation.
+
+The contents of this document are “recommendations”, that is:
+They are more concrete than “principles” (like “write simple code”) because we want them to be clearly action-guiding.
+They are more vague than “rules” (like “never use function X”) because it’s usually impossible to be that precise, and so judgement is still required in applying them.
+
+== Haskell
+
+=== Basic formatting
+
+Use 4 spaces for indentation. Avoid lines over 120 characters. (This is codified in `.editorconfig`.)
+
+Format Haskell source with `stylish-haskell` (config in `.stylish-haskell.yaml`).
+
+==== Reasoning
+
+The indentation is arbitrary. The line length is longer than the “traditional” 80 characters, but anecdotally Haskell code tends toward longer lines, and breaking them up doesn’t help readability much.
+
+`stylish-haskell` is a fairly modest formatter that doesn’t touch too much. At the moment we mostly use it for formatting imports!
+
+In the past we tried `ormolu`, but its https://github.com/tweag/ormolu/issues/641[lack of support for multi-line comments] was a deal-breaker.
+
+=== Write type signatures
+
+Every top-level binding should have a type signature. _Most_ other named bindings (e.g. `let` and `where` bindings) should also have type signatures.
+
+==== Reasoning
+
+Haskell is traditionally lauded for its excellent type inference. People used to make the argument that this brought Haskell closer to dynamically-typed languages in ease of use: you don't have to write type signatures, the compiler will infer it for you.
+
+However, not writing type signatures has a heavy _maintainability_ cost.
+
+.1. Non-local errors
```

Comment:
This seems to format a bit weirdly. Probably these should just be lower-level headings?

Reply:
Actually this is rendering right on github it seems (just like in the google doc). But yeah I can make them headings instead.

Reply:
I get strange stuff where it makes weird collapsible sections or something?